src/EventStore.js

```diff
@@ -75,11 +75,101 @@ class EventStore extends EventEmitter {
+            truncateIndex = true;
+            position--;
+        }
+
+        if (lastEvent && lastEvent.metadata.commitSize && lastEvent.metadata.commitVersion !== lastEvent.metadata.commitSize - 1) {
+            this.emit('unfinished-commit', lastEvent);
+            // commitId = global sequence number at which the commit started
+            this.storage.truncate(lastEvent.metadata.commitId);
+        } else if (truncateIndex) {
+            // The index contained items that are not in the storage file, so truncate it after the last valid event
+            this.storage.truncate(position + 1);
+        }
+        // TODO: Somehow detect unfinished index updates and reindex documents
+        // For this, it needs to be possible to read from a storage without an index
+        // The complexity herein lies in how to correctly interleave reads from different partitions deterministically,
+        // so that they match the order at write-time. With a global sequence number stored in the document meta, this
+        // would be doable relatively easily. Otherwise, we need to:
+        //  - store some other ordering information in the documents, e.g. monotonic timestamp
+        //  - define a deterministic order across partitions, e.g. by ordering by partitionId
+        //  - make sure the order at write-time holds the defined determinism, by guaranteeing that
+        //      given concurrent writes A (partitionId = 1, timestamp = X) and B (partitionId = 2, timestamp = X)
+        //      any index spanning both partitions will always have A come before B
+        //    If the index writer happens to receive B, it hence needs to make sure no write C (partitionId <= 2, timestamp <= X)
+        //    can still occur, before actually appending write B.
+        //    For all partitions part of the index, wait until it is guaranteed that no write to any of those will
+        //    happen with a timestamp <= current write timestamp. This can be guaranteed by delaying the index insert
+        //    by an amount of time that is the maximum delay for a write reaching the index writer. Since there is no guarantee
+        //    for an upper bound this needs to be tackled probabilistically, e.g. by measuring the 99th percentile of latencies.
+        //    Then on any index append, it needs to be checked that the determinism rule is still held. If not, an error needs
+        //    to be thrown, which would mean the according document write needs to be rolled back.
+        //    With this we basically have built a cross-partition transactional scheme, which is contrary to the goals.
+        //    So in contrast, no hard guarantee can be given for repeatable cross-partition reads.
+        this.storage.forEachPartition(partition => {
+            partition.readFi
+        });
+        const indexes = [];
+
+        let leastConsistentEntry = this.storage.index.lastEntry;
+        this.storage.forEachSecondaryIndex(index => {
```

Comment:
Note: Updating secondary indexes is problematic, because they can fall back behind the primary index. However, we do not know if an index has actually fallen behind or if it just didn't match any later documents. This leads to the worst case of one secondary index only matching the first document written, but no other, which would mean the whole storage needs to be reindexed.